### PR TITLE
[LiveComponent] Fix calculateFingerprint with big objects

### PR DESCRIPTION
**Pull request by LunkLoafGrumble**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Tickets       | N/A
| License       | MIT

This pull request will fix an issue for large objects (for example entities with a large relations) with in the FingerprintCalculator

While working with liveComponents I noticed my project getting slower and slower. After some digging I found out that in the calculateFingerprint function the data is serialized. When an entity with relations passes by it will serialize the entity and its child objects. This takes a considerable amount of time and is not needed.

### CODE REVIEW

**Comment by kevinmade** [comment]
I'm having the same issue.

File: src/LiveComponent/src/Util/FingerprintCalculator.php [hunk 1]
```diff
@@ -26,6 +26,21 @@ public function __construct(private string $secret)
 
     public function calculateFingerprint(array $data): string
     {
-        return base64_encode(hash_hmac('sha256', serialize($data), $this->secret, true));
+        return base64_encode(hash_hmac('sha256', serialize($this->normalizeData($data)), $this->secret, true));
+    }
+
+    private function normalizeData(array $data): array
```
Comment (kbond) [inline src/LiveComponent/src/Util/FingerprintCalculator.php]:
```suggestion
    private static function normalizeData(array $data): array
```

**Review by kbond — COMMENTED** [review]
Makes sense to me!

File: src/LiveComponent/src/Util/FingerprintCalculator.php (hunk 1 above)
Comment (LunkLoafGrumble) [reply src/LiveComponent/src/Util/FingerprintCalculator.php]: Done

**Review by kbond — APPROVED** [review]

File: src/LiveComponent/src/Util/FingerprintCalculator.php [hunk 2]
```diff
@@ -26,6 +26,21 @@ public function __construct(private string $secret)
 
     public function calculateFingerprint(array $data): string
     {
-        return base64_encode(hash_hmac('sha256', serialize($data), $this->secret, true));
+        return base64_encode(hash_hmac('sha256', serialize($this->normalizeData($data)), $this->secret, true));
```
Comment (kbond) [inline src/LiveComponent/src/Util/FingerprintCalculator.php]:
```suggestion
        return base64_encode(hash_hmac('sha256', serialize(self::normalizeData($data)), $this->secret, true));
```

File: src/LiveComponent/src/Util/FingerprintCalculator.php [hunk 3]
```diff
@@ -26,6 +26,21 @@ public function __construct(private string $secret)
...
+
+    private static function normalizeData(array $data): array
+    {
+        return array_map(function (mixed $value) {
+            if (\is_object($value)) {
+                return spl_object_id($value);
+            }
+
+            if (\is_array($value)) {
+                return $this->normalizeData($value);
```
Comment (kbond) [inline src/LiveComponent/src/Util/FingerprintCalculator.php]:
```suggestion
                return self::normalizeData($value);
```

**Review by kbond — CHANGES_REQUESTED** [review]

**Review by kbond — APPROVED** [review]

**Comment by weaverryan** [comment]
Great catch on this!

These fingerprints need to be a deterministic string that represents "the input passed into a component" - https://github.com/symfony/ux/blob/e45a8edfcd6d6ab62248d1ffe42ef13e9dcc11b3/src/LiveComponent/src/Util/LiveControllerAttributesCreator.php#L68 - and they need to always return the same string for the same input.

If you pass an object (e.g. a `Post` entity) into a component, the `spl_object_id()` of that might be different the next time the component is re-rendered via an Ajax call, right?

I'm open to ideas on how to fix this. In practice, if we added some special code for entity objects that, if persisted, used their id instead of serializing them, that might solve things. For example, if one of the items in `$data` is a `Post` entity, then instead of serializing that or `spl_object_id()`, we use `App\Entity\Post:{THE_ID}`.

**Comment by kbond** [comment]
> These fingerprints need to be a deterministic string that represents

Oops, didn't realize this. Could `spl_object_hash()` be used? I don't know if this would still have a perf problem.

**Comment by weaverryan** [comment]
I think `spl_object_hash()` has the same problem - not 100% certain on that, but pretty sure.

**Comment by kbond** [comment]
You're right. From php.net:

> When an object is destroyed, its hash may be reused for other objects.

**Comment by kbond** [comment]
> These fingerprints need to be a deterministic string that represents "the input passed into a component" -

We should have a test to prove this behaviour.

**Comment by kevinmade** [comment]
Perhaps it's an idea to add `if (\is_object($value) && method_exists($value, 'getId')) {
                return $value->getId();
            }` instead of `\is_object($value)`.

File: src/LiveComponent/src/Util/FingerprintCalculator.php [hunk 4]
```diff
@@ -20,12 +25,29 @@
...
 
     public function calculateFingerprint(array $data): string
     {
-        return base64_encode(hash_hmac('sha256', serialize($data), $this->secret, true));
+        return base64_encode(hash_hmac('sha256', serialize($this->normalizeData($data)), $this->secret, true));
+    }
+
+    private function normalizeData(array $data): array
+    {
+        return array_map(function (mixed $value) {
```
Comment (kbond) [inline src/LiveComponent/src/Util/FingerprintCalculator.php]: Can we just call `return $this->objectNormalizer->normalize($value, context: [LiveComponentHydrator::LIVE_CONTEXT => true]);` (or inline above)?

**Review by kbond — CHANGES_REQUESTED** [review]
This is probably the best solution.

@LunkLoafGrumble does this method improve the speed in your app?

Could you add a basic integration test for `FingerprintCalculator` that proves the generated fingerprint is the same two different object instances that are equal?

File: src/LiveComponent/src/Util/FingerprintCalculator.php (hunk 4 above)
Comment (LunkLoafGrumble) [reply src/LiveComponent/src/Util/FingerprintCalculator.php]: Good suggestion. changed it.

**Comment by LunkLoafGrumble** [comment]
Using objectNormalizer greatly improves the performance over just serializing the object.

File: src/LiveComponent/tests/Integration/Util/FingerprintCalculatorTest.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    public function testFingerprintEqual()
+    {
+        $fingerprintCalculator = self::getContainer()->get('ux.live_component.fingerprint_calculator');
+
+        $entityOne = create(Entity1::class)->object();
+
+        $entityTwo = clone $entityOne;
+
+        self::assertNotEquals(
```
Comment (kbond) [inline src/LiveComponent/tests/Integration/Util/FingerprintCalculatorTest.php]:
```suggestion
        self::assertNotSame(
```

File: src/LiveComponent/tests/Integration/Util/FingerprintCalculatorTest.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            spl_object_id($entityOne),
+            spl_object_id($entityTwo)
+        );
+
+        self::assertEquals(
```
Comment (kbond) [inline src/LiveComponent/tests/Integration/Util/FingerprintCalculatorTest.php]:
```suggestion
        self::assertSame(
```

File: src/LiveComponent/tests/Integration/Util/FingerprintCalculatorTest.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    public function testFingerprintNotEqual()
+    {
+        $fingerprintCalculator = self::getContainer()->get('ux.live_component.fingerprint_calculator');
+
+        $entityOne = create(Entity1::class)->object();
+
+        $entityTwo = create(Entity1::class)->object();
+
+        self::assertNotEquals(
```
Comment (kbond) [inline src/LiveComponent/tests/Integration/Util/FingerprintCalculatorTest.php]:
```suggestion
        self::assertNotSame(
```

**Review by kbond — CHANGES_REQUESTED** [review]
A few remaining minor things, then good to go by me!

**Review by kbond — APPROVED** [review]

**Comment by weaverryan** [comment]
Let's give this version a try - thanks @LunkLoafGrumble!

**Comment by weaverryan** [comment]
FYI - I'm reverting this in #832. But I'm hoping that it's not a problem anymore because usually zero LiveProps are used to calculate the fingerprint. But please tell me if you still have issues.

**Pull request merged**